packages/cli/cloud/src/environment/link/action.ts

```diff
@@ -0,0 +1,140 @@
+}
+
+interface LinkEnvironmentInput extends Answers {
+  targetEnvironment: string;
+}
+
+type Environment = {
+  name: string;
+};
+
+type EnvironmentsList = Environment[];
+
+export default async (ctx: CLIContext) => {
+  const { getValidToken } = await tokenServiceFactory(ctx);
+  const token = await getValidToken(ctx, promptLogin);
+  const { logger } = ctx;
+
+  if (!token) {
+    return;
+  }
+
+  const project = await getLocalProject(ctx);
+
+  if (!project) {
+    logger.debug(`No valid local project configuration was found.`);
+    return;
+  }
+
+  const cloudApiService = await cloudApiFactory(ctx, token);
+  const environments = await getEnvironmentsList(ctx, cloudApiService, project);
+
+  if (!environments) {
+    return;
+  }
+
+  const answer: LinkEnvironmentAnswer | null = await promptUserForEnvironment(ctx, environments);
+
+  if (!answer) {
+    return;
+  }
+
+  await trackEvent(ctx, cloudApiService, 'willLinkEnvironment', {
+    projectName: project.name,
+    environmentName: answer.targetEnvironment,
+  });
+
+  try {
+    await local.addEnvironment(answer.targetEnvironment);
+  } catch (e) {
+    await trackEvent(ctx, cloudApiService, 'didNotLinkEnvironment', {
+      projectName: project.name,
+      environmentName: answer.targetEnvironment,
+    });
+    logger.debug('Failed to link environment', e);
+    logger.error('An error occurred while trying to link the environment.');
+    process.exit(1);
+  }
+
+  logger.log(
+    `You have successfully linked your project to ${chalk.cyan(answer.targetEnvironment)}, on ${chalk.cyan(project.displayName)}. You are now able to deploy your project.`
+  );
+  await trackEvent(ctx, cloudApiService, 'didLinkEnvironment', {
+    projectName: project.name,
+    environmentName: answer.targetEnvironment,
+  });
+};
+
+async function promptUserForEnvironment(
+  ctx: CLIContext,
+  environments: EnvironmentsList
+): Promise<LinkEnvironmentAnswer | null> {
+  const { logger } = ctx;
+
+  try {
+    const answer: LinkEnvironmentInput = await inquirer.prompt([
+      {
+        type: 'list',
+        name: 'targetEnvironment',
+        message: 'Which environment do you want to link?',
+        choices: [...environments, { name: chalk.grey(`(${QUIT_OPTION})`), value: null }],
+      },
+    ]);
+
+    if (!answer.targetEnvironment) {
+      return null;
+    }
+
+    return answer;
+  } catch (e) {
+    logger.debug('Failed to get user input', e);
+    logger.error('An error occurred while trying to get your environment selection.');
+    return null;
+  }
+}
+
+async function getEnvironmentsList(
+  ctx: CLIContext,
+  cloudApiService: CloudApiService,
+  project: ProjectInput
+) {
+  const spinner = ctx.logger.spinner('Fetching environments...\n').start();
+
+  try {
+    const {
+      data: { data: environmentsList },
+    } = await cloudApiService.listLinkEnvironments({ name: project.name });
+    spinner.succeed();
+
+    return (environmentsList as unknown as Environment[])
```

Comment:
Use type inference to avoid casting. Try to add this before the line:

```
    if (!Array.isArray(environmentsList)) {
      return [] or undefined or null, or throw error
    }
```